src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+    SuccessOrExit(err);
+
+    err = apContext->mWriter.StartContainer(ContextTag(EventDataElement::kCsTag_Data), TLV::kTLVType_Structure, dataContainerType);
+    SuccessOrExit(err);
+    // Callback to write the EventData
+    err = apDelegate->WriteEvent(apContext->mWriter);
+    SuccessOrExit(err);
+
+    err = apContext->mWriter.EndContainer(dataContainerType);
+    SuccessOrExit(err);
+
+    eventDataElementBuilder.EndOfEventDataElement();
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.Finalize();
+    SuccessOrExit(err);
+
+    apContext->mFirst = false;
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        apContext->mWriter = checkpoint;
+    }
+    else
+    {
+        // update these variables since ConstructEvent can be used to track the
+        // state of a set of events over multiple calls.
+        apContext->mCurrentEventNumber++;
+        if (apContext->mCurrentSystemTime.mType == Timestamp::Type::kSystem)
+        {
+            apContext->mCurrentSystemTime = apOptions->mTimestamp;
+        }
+    }
+    return err;
+}
+
+void EventManagement::CreateEventManagement(Messaging::ExchangeManager * apExchangeManager, int aNumBuffers,
+                                            CircularEventBuffer * apCircularEventBuffer,
+                                            const LogStorageResources * const apLogStorageResources)
+{
+
+    sInstance.Init(apExchangeManager, aNumBuffers, apCircularEventBuffer, apLogStorageResources);
+    static_assert(std::is_trivially_destructible<EventManagement>::value, "EventManagement must be trivially destructible");
+}
+
+/**
+ * @brief Perform any actions we need to on shutdown.
+ */
+void EventManagement::DestroyEventManagement()
+{
+    Platform::CriticalSectionEnter();
+    sInstance.mState        = EventManagementStates::Shutdown;
+    sInstance.mpEventBuffer = nullptr;
+    sInstance.mpExchangeMgr = nullptr;
+    Platform::CriticalSectionExit();
+}
+
+EventNumber CircularEventBuffer::VendEventNumber()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Assign event Number to the buffer's counter's value.
+    mLastEventNumber = static_cast<EventNumber>(mpEventNumberCounter->GetValue());
+
+    // Now advance the counter.
+    err = mpEventNumberCounter->Advance();
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(EventLogging, "%s Advance() for priority %d failed with %d", __FUNCTION__, mPriority, err);
+    }
+
+    return mLastEventNumber;
+}
+
+EventNumber EventManagement::GetLastEventNumber(PriorityLevel aPriority)
+{
+    return GetPriorityBuffer(aPriority)->mLastEventNumber;
+}
+
+EventNumber EventManagement::GetFirstEventNumber(PriorityLevel aPriority)
+{
+    return GetPriorityBuffer(aPriority)->mFirstEventNumber;
+}
+
+CircularEventBuffer * EventManagement::GetPriorityBuffer(PriorityLevel aPriority) const
+{
+    CircularEventBuffer * buf = mpEventBuffer;
+    while (!buf->IsFinalDestinationForPriority(aPriority))
```

Comment:
GetNextCircularEventBuffer can be nullptr below.

Consider adding an assert that buf is never nullptr with the comment that code guarantees that every PriorityLevel has a buffer destination.